packages/dd-trace/src/opentelemetry/metrics/otlp_transformer.js

```diff
@@ -0,0 +1,286 @@
+ *
+ * @class OtlpTransformer
+ * @extends OtlpTransformerBase
+ */
+class OtlpTransformer extends OtlpTransformerBase {
+  /**
+   * Creates a new OtlpTransformer instance.
+   *
+   * @param {import('@opentelemetry/api').Attributes} resourceAttributes - Resource attributes
+   * @param {string} protocol - OTLP protocol (http/protobuf or http/json)
+   */
+  constructor (resourceAttributes, protocol) {
+    super(resourceAttributes, protocol, 'metrics')
+  }
+
+  /**
+   * Transforms metrics to OTLP format based on the configured protocol.
+   * @param {Array} metrics - Array of metric data to transform
+   * @returns {Buffer} Transformed metrics in the appropriate format
+   */
+  transformMetrics (metrics) {
+    if (this.protocol === 'http/json') {
+      return this.#transformToJson(metrics)
+    }
+    return this.#transformToProtobuf(metrics)
+  }
+
+  /**
+   * Transforms metrics to protobuf format.
+   * @param {Array} metrics - Array of metrics to transform
+   * @returns {Buffer} Protobuf-encoded metrics
+   * @private
+   */
+  #transformToProtobuf (metrics) {
+    const { protoMetricsService } = getProtobufTypes()
+
+    const metricsData = {
+      resourceMetrics: [{
+        resource: this._transformResource(),
+        scopeMetrics: this.#transformScope(metrics),
+        schemaUrl: ''
+      }]
+    }
+
+    return this._serializeToProtobuf(protoMetricsService, metricsData)
+  }
+
+  /**
+   * Transforms metrics to JSON format.
+   * @param {Array} metrics - Array of metrics to transform
+   * @returns {Buffer} JSON-encoded metrics
+   * @private
+   */
+  #transformToJson (metrics) {
+    const metricsData = {
+      resourceMetrics: [{
+        resource: this._transformResource(),
+        scopeMetrics: this.#transformScope(metrics, true)
+      }]
+    }
+    return this._serializeToJson(metricsData)
+  }
+
+  /**
+   * Creates scope metrics grouped by instrumentation scope.
+   * @param {Array} metrics - Array of metrics to transform
+   * @param {boolean} isJson - Whether to format for JSON output
+   * @returns {Object[]} Array of scope metric objects
+   * @private
+   */
+  #transformScope (metrics, isJson = false) {
+    const groupedMetrics = this._groupByInstrumentationScope(metrics)
+    const scopeMetrics = []
+
+    for (const [, metricsInScope] of groupedMetrics) {
+      // Get the first metric to extract the instrumentation scope details
+      const firstMetric = metricsInScope[0]
+      const instrumentationScope = firstMetric.instrumentationScope || {}
+      const { name = '', version = '', schemaUrl = '', attributes = {} } = instrumentationScope
+
+      const scope = {
+        name,
+        version,
+        droppedAttributesCount: 0
+      }
+
+      // Add attributes if present
+      if (attributes && Object.keys(attributes).length > 0) {
+        scope.attributes = isJson ? this._attributesToJson(attributes) : this._transformAttributes(attributes)
+      }
```

Comment:
Both of these methods are actually going to create an empty array for an empty object.
So the empty case has the same overhead when calling the method directly. When the object has entries, it would now create an additional array of all keys. To prevent this, we can just get the return type of the method and check if the return is empty or not. That way there is no additional overhead anymore.

```suggestion
      if (attributes) {
        const transformed = isJson ? this._attributesToJson(attributes) : this._transformAttributes(attributes)
        if (transformed.length) {
          scope.attributes = transformed
        }
      }
```